### ISSUE TITLE
Contracts surrounded by 2 empty lines functions by 1

### ISSUE THREAD

**Issue by Janther** [prettier-solidity/prettier-plugin-solidity#143]
This will require some redoing of the `printPreservingEmptyLines` function.

After Prettier, code should look like this:

```Solidity
pragma 0.5.0;


library Math {
    function add(uint256 a, uint256 b) internal pure returns (uint256) {
        uint256 c = a + b;
        require(c >= a, "SafeMath: addition overflow");

        return c;
    }
}


interface IERC20 {
    function balanceOf(address account) external view returns (uint256);
    function transfer(address recipient, uint256 amount) external returns (bool);
    event Transfer(address indexed from, address indexed to, uint256 value);
}


contract Token is IERC20 {
    using SafeMath for uint256;

    mapping (address => uint256) private _balances;

    function balanceOf(address account) public view returns (uint256) {
        return _balances[account];
    }

    function transfer(address recipient, uint256 amount) public returns (bool) {
        _transfer(msg.sender, recipient, amount);
        return true;
    }

    function _transfer(address sender, address recipient, uint256 amount) internal {
        require(sender != address(0), "ERC20: transfer from the zero address");
        require(recipient != address(0), "ERC20: transfer to the zero address");

        _balances[sender] = _balances[sender].sub(amount);
        _balances[recipient] = _balances[recipient].add(amount);
        emit Transfer(sender, recipient, amount);
    }
}
```

**Comment by fvictorio** [comment]
Is this in the style guide? Also, do you have some info on how much this is used in the industry?

**Comment by Janther** [comment]
It's in one of the first sections of the Styleguide. [Blank Lines](https://solidity.readthedocs.io/en/v0.5.9/style-guide.html#blank-lines)

The linter [Solhint](https://github.com/protofire/solhint) enforces it.

Is very seldom that a file contains more than one contract.
This is mostly seen in flat files to verify contracts in Etherscan.
however I believe that we are not enforcing the separation between functions, only allowing it if the developer put the blank line before.

**Comment by fvictorio** [comment]
In JS prettier, the philosophy is to never add blank lines. The only thing they do is to merge multiple blank lines into one. That's why we don't add blank lines between functions. I like that approach, but it's not written in stone.

If you want to enforce blank lines between functions, you can use a linter rule. I don't remember right now if solhint has something like that, but it could be added. Then, if you fix that, prettier won't mess with that change.

But the "two blank lines between contracts" is a little harder. If you are using both solhint and prettier, and solhint forces you to add them, prettier will later merge them. So, the options here are:

- Make prettier enforce this, which is weird because it doesn't do it for functions.
- Modify prettier's behavior so that it only merges consecutive blank lines if there are 3 or more of them (and it merges into two). I don't love this idea, but it could be a compromise. You could also argue for doing this only at the top level, and merge into one blank line inside contracts, but this might be unnecessary complexity.

(For the record, my intention is to eventually remove as many styles rules as possible from solhint and delegate that to prettier, but there are scenarios like this one where it isn't clear who should be the responsible.)

**Comment by Janther** [comment]
This one requires a lot of thinking and testing to do properly and the benefit is slim, so it's very low in my priority list, but I want us to be aware of it.

**Comment by Janther** [comment]
Fixed in #217

**Comment by PaulRBerg** [comment]
> The linter Solhint enforces it.

Not the case anymore with solhint v3 and above. I can lint a contract with only one line between the import statement and the contract name and no error or warning is thrown.
